crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,823 @@
+# Copyright (c) Mysten Labs, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+# Open Questions
+# --------------
+# - Rules for nullabilit/non-nullability
+# - Pattern for multi-gets
+# - Encoding RPC errors
+# - `type` (vs `scalar`) for Move Types
+# - Pattern for upcasting and downcasting
+
+# Check Completeness Against Existing APIs
+# -------------
+# DONE Read
+# DONE - Objects
+# DONE - Transactions
+# DONE - Checkpoints
+# DONE - Events
+# DONE - ProtocolConfig
+# DONE - Chain Identifier
+# DONE Coin
+# DONE Extended
+# DONE Governance
+# DONE Indexer
+# DONE Move Utils
+# TODO Write
+
+# DONE Port designs from Notion
+# DONE - Display
+# DONE - SuiNS
+# DONE Data Consistency API
+# TODO Flesh out Connections
+# TODO - TotalTransactionBlocks for the PTB connection
+
+schema {
+  query: Query
+  subscription: Subscription
+}
+
+type Query {
+  # First four bytes of the network's genesis checkpoint digest
+  # (uniquely defines the network)
+  chainIdentifier: String!
+
+  # First and last checkpoint that the RPC can answer consistent
+  # queries for.
+  consistentRange: ConsistentRange
+
+  owner(address: SuiAddress!): Owner
+  object(address: SuiAddress!, version: BigInt): Object
+  address(address: SuiAddress!): Address
+
+  epoch(epochId: Int): Epoch
+  protocolConfig(protocolVersion: Int): ProtocolConfig
+
+  checkpoint(filter: CheckpointFilter): Checkpoint
+  checkpoints(order: Order!): CheckpointConnection
+
+  transactionBlock(digest: Bytes!): TransactionBlock
+  transactionBlocks: TransactionBlockConnection
+
+  coinMetadata(coinType: String!): CoinMetadata
+
+  queryObjects(query: ObjectQuery): ObjectQueryConnection
+  queryEvents(filter: EventFilter, order: Order!): EventQueryConnection
+  queryTransactionBlocks(
+    filter: TransactionBlockFilter,
+    order: Order!,
+  ): TransactionBlockQueryConnection
+
+  resolveNameServiceAddress(name: String!): Address
+
+  networkMetrics: NetworkMetrics
+  moveCallMetrics: MoveCallMetrics
+  checkpointAddressMetrics(checkpoint: BigInt): AddressMetric
+  allEpochAddressMetrics(order: Order!): AddressMetricConnection
+}
+
+type Subscription {
+  events(filter: EventFilter): Event
+  transactions(filter: TransactionBlockFilter): TransactionBlock
+}
+
+# String containing 32B hex-encoded address
+scalar SuiAddress
+
+# String representation of an arbitrary width, possibly signed integer
+scalar BigInt
+
+# String containing Base58-encoded binary data.
+scalar Bytes
+
+# Filter either by the digest, or the sequence number, or neither, to
+# get the latest checkpoint.
+input CheckpointFilter {
+  digest: Bytes
+  sequenceNumber: BigInt
+}
+
+enum Order { ASC, DESC }
+
+input ObjectQuery {
+  # one of these
+  any: [ObjectQuery]
+  all: [ObjectQuery]
+  not: ObjectQuery
+
+  # (x)or any of these
+
+  # Cascading (type requires module requires package)
+  package: SuiAddress
+  module: String
+  type: String
+
+  owner: SuiAddress
+  objectId: SuiAddress
```

Comment:
one thing that is slightly inconvenient about this style is the fairly common pattern of

`any: [{ objectId: id1 }, { objectId: id1 }]`, it might be worth also providing a list of addresses, or maybe only supporting lists of addresses rather than a single address.  In practice there is no real reason to use this query if you only want a single address since you could just load the object directly

Reply:
I started going down this road a little bit, but it quickly unravelled because you probably also need to support object keys (id + version), and then there's a question of whether you have an 

```
input ObjectKey {
    id: SuiAddress!
    version: U64
}
```

and then accept a list of those, or to accept separate lists of keys and bare IDs... Thoughts on waiting to see how the API is used to see what the most helpful conveniences are?  Seems easier to evolve that part of the schema in tandem with the client.